File: Medium/2026-08-29-2948-Make-Lexicographically-Smallest-Array-by-Swapping-Elements/solution.py
class Solution:
    def smallestArray(self, nums: List[int], limit: int) -> List[int]:
        n = len(nums)
        parent = list(range(n))

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            rootX = find(x)
            rootY = find(y)
            if rootX != rootY:
                parent[rootY] = rootX

        for i in range(n):
            for j in range(i + 1, n):
                if abs(nums[i] - nums[j]) <= limit:
                    union(i, j)

        groups = {}
        for i in range(n):
            root = find(i)
            if root not in groups:
                groups[root] = []
            groups[root].append(nums[i])

        result = [0] * n
        for indices in groups.values():
            indices.sort()
            for idx, value in zip(sorted(groups.keys()), indices):
                result[idx] = value

        return result

File: Medium/2026-08-29-2948-Make-Lexicographically-Smallest-Array-by-Swapping-Elements/solution.js
class UnionFind {
    constructor(n) {
        this.parent = Array.from({ length: n }, (_, i) => i);
    }

    find(x) {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
    }

    union(x, y) {
        const rootX = this.find(x);
        const rootY = this.find(y);
        if (rootX !== rootY) {
            this.parent[rootY] = rootX;
        }
    }
}

var smallestArray = function(nums, limit) {
    const n = nums.length;
    const uf = new UnionFind(n);

    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
            if (Math.abs(nums[i] - nums[j]) <= limit) {
                uf.union(i, j);
            }
        }
    }

    const groups = {};
    for (let i = 0; i < n; i++) {
        const root = uf.find(i);
        if (!groups[root]) {
            groups[root] = [];
        }
        groups[root].push(nums[i]);
    }

    const result = new Array(n);
    for (const group of Object.values(groups)) {
        group.sort((a, b) => a - b);
        for (let i = 0; i < group.length; i++) {
            const index = nums.indexOf(group[i]);
            result[index] = group[i];
            nums[index] = null; // Mark as used
        }
    }

    return result;
};